limbo/testcases/rfc5280.py

```diff
@@ -611,3 +612,425 @@ def leaf_ku_keycertsign(builder: Builder) -> None:
+
+    ```
+    root -> leaf
+    ```
+
+    The root contains a NameConstraints extension with a permitted DirectoryName
+    of "CN=foo". This should not match the child's DirectoryName of "CN=not-foo".
+    """
+    root = builder.root_ca(
+        name_constraints=ext(
+            x509.NameConstraints(
+                [x509.DirectoryName(x509.Name.from_rfc4514_string("CN=foo"))], None
+            ),
+            critical=True,
+        )
+    )
+    leaf = builder.leaf_cert(
+        root,
+        subject=x509.Name.from_rfc4514_string("CN=not-foo"),
+        san=ext(
+            x509.SubjectAlternativeName(
+                [x509.DirectoryName(x509.Name.from_rfc4514_string("CN=not-foo"))]
+            ),
+            critical=False,
+        ),
+    )
+
+    builder = builder.client_validation()
+    builder.trusted_certs(root).peer_certificate(leaf).fails()
+
+
+@testcase
+def ca_nameconstraints_excluded_dn_match(builder: Builder) -> None:
+    """
+    Produces the following **invalid** chain:
+
+    ```
+    root -> leaf
+    ```
+
+    The root contains a NameConstraints extension with an excluded DirectoryName
+    of "CN=foo", matching the leaf's SubjectAlternativeName.
+    """
+    root = builder.root_ca(
+        name_constraints=ext(
+            x509.NameConstraints(
+                None,
+                [x509.DirectoryName(x509.Name.from_rfc4514_string("CN=foo"))],
+            ),
+            critical=True,
+        )
+    )
+    leaf = builder.leaf_cert(
+        root,
+        subject=x509.Name.from_rfc4514_string("CN=foo"),
+        san=ext(
+            x509.SubjectAlternativeName(
+                [x509.DirectoryName(x509.Name.from_rfc4514_string("CN=foo"))]
+            ),
+            critical=False,
+        ),
+    )
+
+    builder = builder.client_validation()
+    builder.trusted_certs(root).peer_certificate(leaf).fails()
+
+
+@testcase
+def ca_nameconstraints_permitted_dn_match(builder: Builder) -> None:
+    """
+    Produces the following **valid** chain:
+
+    ```
+    root -> leaf
+    ```
+
+    The root contains a NameConstraints extension with a permitted DirectoryName
+    of "CN=foo", matching the leaf's SubjectAlternativeName.
+    """
+    root = builder.root_ca(
+        name_constraints=ext(
+            x509.NameConstraints(
+                [x509.DirectoryName(x509.Name.from_rfc4514_string("CN=foo"))], None
+            ),
+            critical=True,
+        )
+    )
+    leaf = builder.leaf_cert(
+        root,
+        subject=x509.Name.from_rfc4514_string("CN=foo"),
+        san=ext(
+            x509.SubjectAlternativeName(
+                [x509.DirectoryName(x509.Name.from_rfc4514_string("CN=foo"))]
+            ),
+            critical=False,
+        ),
+    )
+
+    builder = builder.client_validation()
+    builder.trusted_certs(root).peer_certificate(leaf).succeeds()
+
+
+@testcase
+def ca_nameconstraints_permitted_dn_match_sub_mismatch(builder: Builder) -> None:
```

Comment:
Not sure if this makes it clearer or not.

```suggestion
def ca_nameconstraints_permitted_dn_match_subject_san_mismatch(builder: Builder) -> None:
```